### PR TITLE
Fixed undo issue given in #2545

### PR DESCRIPTION
**Pull request by Chillee**
Thanks to @Diadlo for this repro. https://github.com/VSCodeVim/Vim/issues/2545.

Basically, we don't always want to create a `prevHandler` even if it doesn't exist. To be honest, this code is messy and I'd like to rewrite it entirely, if it wasn't for the fact that this kind of frail code is hard to test.

This should fix many of the reports given in https://github.com/VSCodeVim/Vim/issues/2007. However, the first reports of the issue started before this repro was introduced. That suggests to me that there's actually 2 separate bugs. One where the file gets replaced by another file (issue 1), and one where the file actually drops back to an earlier undo state (issue 2).

Issue 1 was introduced around January 12th in this seemingly innocuous commit: c05006620562db74c43aa73f7a9adb14073ebcb6 .

The first report of this (replaced with another file is here): https://github.com/VSCodeVim/Vim/issues/2007#issuecomment-361799959 Many of the following comments were also expressing this issue.

However, the first report of issue 2 were given here: https://github.com/VSCodeVim/Vim/issues/928 The issue is slightly confused by the fact that the initial bug that @octref reports is a much more innocuous "Certain VSCode actions aren't given their own undo stop", while issue 2 is really "VSCodeVim is wiping out a ton of work". The first real report of issue 2 appears to be https://github.com/VSCodeVim/Vim/issues/928#issuecomment-317941397, on July 26th. I believe that this bug was likely introduced by this commit: https://github.com/VSCodeVim/Vim/pull/1629 on May 3rd.

This commit fixes issue 1. More work is needed to locate the cause of issue 2.

@jpoon @johnfn @xconverge @rebornix

### CODE REVIEW

**Comment by jpoon** [comment]
> Basically, we don't always want to create a prevHandler even if it doesn't exist. 

Why?

**Comment by Chillee** [comment]
@jpoon When we close a file, we delete its handler from `ModeHandlerMap`. When we were calling `ModeHandlerMap.getOrCreate` in `getAndUpdateModeHandler`, `prevHandler` did not exist at that step. So, it would create a new handler for the file we just closed. If you take a look at the code for creating a new mode handler, it creates it out of the current file.

This caused us to initialize the undo history of the closed file with the text of the current file. Then, when you reopened that file, pressing undo would set you to the text of the other file.

So, in this repro: https://github.com/VSCodeVim/Vim/issues/2545
1. We have the correct state for document A.
2. We have the correct state for document A & B.
3. Closing document B first deletes the current `modeHandler` we have, but `getOrCreate` creates another modeHandler initialized with the vimState of document A. So, we have the correct state for document A but the wrong one for document B. Only document A is currently open.
4. Opening this file and changing the contents is treated as an undoable change.
5-6. Undoing overwrites the contents of document B with document A.

File: extension.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ let extensionContext: vscode.ExtensionContext;
 let previousActiveEditorId: EditorIdentity = new EditorIdentity();
 
 export async function getAndUpdateModeHandler(): Promise<ModeHandler> {
-  const [prevHandler] = await ModeHandlerMap.getOrCreate(previousActiveEditorId.toString());
```
Comment (jpoon) [inline extension.ts]: Maybe we can change this to a `create` function, but I can do that refactor in a later commit.

File: src/mode/modeHandlerMap.ts [hunk 2]
```diff
@@ -16,6 +16,11 @@ class ModeHandlerMapImpl {
     return [modeHandler, isNew];
   }
 
+  async get(key: string): Promise<ModeHandler | null> {
+    let modeHandler = this.modeHandlerMap[key];
```
Comment (jpoon) [inline src/mode/modeHandlerMap.ts]: Can probably do this all in one line.

**Comment by johnfn** [comment]
Oh man what a hero!

**Comment by johnfn** [comment]
I agree with your conclusion, too. The ModeHandler stuff is a mess and has been responsible for more bugs than I could count. Worst part is, the biggest benefit of ModeHandler is per-file state, and I can't even remember why we need that. Nearly everything is per-project.

**Comment by jcjolley** [comment]
@Chillee, you're amazing.

**Comment by jpoon** [comment]
@Chillee ok to merge?

**Comment by Chillee** [comment]
Yep!

**Comment by jpoon** [comment]
I'll cut a release for it.

**Comment by rifotu** [comment]
great guys!
thanks a lot :)

**Pull request merged**